indra/tools/fix_invalidities.py

```diff
@@ -0,0 +1,155 @@
+    for k, v in copy.deepcopy(ev.text_refs).items():
+        if v is None:
+            ev.text_refs.pop(k, None)
+        elif not k.isupper():
+            ev.text_refs.pop(k)
+            ev.text_refs[k.upper()] = v
+
+    if ev.pmid and not re.match(text_ref_patterns['PMID'], ev.pmid):
+        ev.pmid = None
+    if ev.text_refs.get('PMID') and not re.match(text_ref_patterns['PMID'],
+                                                 ev.text_refs['PMID']):
+        ev.text_refs.pop('PMID', None)
+
+    if ev.pmid is None and ev.text_refs.get('PMID') is not None:
+        ev.pmid = ev.text_refs['PMID']
+    elif ev.text_refs.get('PMID') is None and ev.pmid is not None:
+        ev.text_refs['PMID'] = ev.pmid
+
+    if 'DOI' in ev.text_refs and not re.match(text_ref_patterns['DOI'],
+                                              ev.text_refs['DOI']):
+        ev.text_refs.pop('DOI', None)
+    if 'PMC' in ev.text_refs and not re.match(text_ref_patterns['PMC'],
+                                              ev.text_refs['PMC']):
+        ev.text_refs.pop('PMC', None)
+
+    if ev.context is not None:
+        fix_invalidities_context(ev.context)
+
+
+def fix_invalidities_agent(agent: Agent):
+    """Fix invalidities of a single INDRA Agent in place."""
+    agent.db_refs = fix_invalidities_db_refs(agent.db_refs)
+
+
+def fix_invalidities_db_refs(db_refs: Mapping[str, str]) -> Mapping[str, str]:
+    """Return a fixed version of a db_refs grounding dict."""
+    if 'PUBCHEM' in db_refs and \
+            db_refs['PUBCHEM'].startswith('CID'):
+        db_refs['PUBCHEM'] = \
+            db_refs['PUBCHEM'].replace('CID:', '').strip()
+
+    db_refs = {k: v for k, v in db_refs.items()
+               if v is not None}
+
+    for k, v in copy.deepcopy(db_refs).items():
+        if k == 'CHEMBL' and not v.startswith('CHEMBL'):
+            db_refs[k] = 'CHEMBL%s' % v
+        elif k == 'ECCODE':
+            db_refs['ECCODE'] = db_refs['ECCODE'].replace('.-', '')
+        elif k == 'UNIPROT':
+            db_refs.pop(k)
+            # This is really a location
+            if v.startswith('SL-'):
+                db_refs['UPLOC'] = v
+            # Otherwise we just fix the invalid key
+            else:
+                db_refs['UP'] = v
+        elif k == 'UP':
+            # There are cases where this is an empty string
+            if not v.strip():
+                db_refs.pop('UP', None)
+            # Sometimes we have two IDs separated by a comma
+            if ',' in v:
+                db_refs['UP'] = v.split(',')[0]
+            if v.startswith('SL-'):
+                db_refs['UPLOC'] = db_refs.pop('UP')
+        elif k == 'UAZ':
+            db_refs.pop('UAZ')
+            if v.startswith('CVCL'):
+                db_refs['CVCL'] = v
+        elif k == 'TAXONOMY' and v == '-1':
+            db_refs.pop('TAXONOMY', None)
+        elif k == 'LINCS' and re.match(r'\d+-\d+', v):
```

Comment:
sorry for the late comment but it might be good to pre-compile these regular expressions so it doesn't have to compile it inline every time